src/invite-api.js

```diff
@@ -1,236 +1,440 @@
+   * connect to the invitor's device.
+   *
+   * If the invite is accepted and you had other invites to the same project,
+   * those invites are removed, and the invitors are told that you're already
+   * part of this project.
+   *
+   * @param {Pick<Invite, 'inviteId'>} invite
    * @returns {Promise<void>}
    */
-  async accept(projectPublicId) {
-    const projectId = this.#getProjectIdFromPublicId(projectPublicId)
+  async accept({ inviteId: inviteIdString }) {
+    const inviteId = Buffer.from(inviteIdString, 'hex')
+
+    const pendingInvite = this.#pendingInvites.getByInviteId(inviteId)
+    if (!pendingInvite) {
+      throw new Error(`Cannot find invite ID ${inviteIdString}`)
+    }
+
+    const { peerId, invite } = pendingInvite
+    const { projectPublicId } = invite
 
-    const isAlreadyMember = await this.#isMember(projectId)
-    const peersToRespondTo = this.#peersToRespondTo.get(projectId)
+    const removePendingInvite = () => {
+      const didDelete = this.#pendingInvites.deleteByInviteId(inviteId)
+      if (didDelete) this.emit('invite-removed', internalToExternal(invite))
+    }
 
+    // This is probably impossible in the UI, but it's theoretically possible
+    // to join a project while an invite is pending, so we need to check this.
+    const isAlreadyMember = this.#isMember(projectPublicId)
     if (isAlreadyMember) {
-      for (const peerId of peersToRespondTo) {
-        await this.#sendAlreadyResponse({ peerId, projectId })
+      const pendingInvitesDeleted =
+        this.#pendingInvites.deleteByProjectPublicId(projectPublicId)
+      for (const pendingInvite of pendingInvitesDeleted) {
+        this.#sendAlreadyResponse({
+          peerId: pendingInvite.peerId,
+          inviteId: pendingInvite.invite.inviteId,
+        })
+        this.emit('invite-removed', internalToExternal(pendingInvite.invite))
       }
       return
     }
 
-    const pendingInvite = this.#pendingInvites.get(projectId)
-
-    if (!pendingInvite) {
-      throw new Error(
-        `Cannot find invite for project with ID ${projectPublicId}`
-      )
+    const { promise: projectDetailsPromise, resolve: gotProjectDetails } =
+      /** @type {typeof promiseWithResolvers<ProjectJoinDetails>} */
+      (promiseWithResolvers)()
+    /**
+     * @param {string} projectDetailsPeerId
+     * @param {ProjectJoinDetails} details
+     */
+    const onProjectDetails = (projectDetailsPeerId, details) => {
+      const isDetailsForThisInvite =
+        // This peer ID check is probably superfluous because the invite ID
+        // should be impossible to guess, but is easy to add for correctness.
+        timingSafeEqual(projectDetailsPeerId, peerId) &&
+        timingSafeEqual(inviteId, details.inviteId)
+      if (!isDetailsForThisInvite) return
+
+      gotProjectDetails(details)
+      this.rpc.off('got-project-details', onProjectDetails)
     }
+    this.rpc.on('got-project-details', onProjectDetails)
```

Comment:
I don't think this needs to be a deferred promise, and I've tried to avoid them when I can, purely based on [this random advice](https://github.com/petkaantonov/bluebird/wiki/Promise-anti-patterns#the-deferred-anti-pattern) that I don't know if it's correct. But I think this is maybe easier to read as a regular promise, something like:

```js
const projectDetailsPromise = new Promise(resolve => {
  const { rpc } = this
  rpc.on('got-project-defailts', function onProjectDetails(projectDetailsPeerId, details) {
    const isDetailsForThisInvite =
      // This peer ID check is probably superfluous because the invite ID
      // should be impossible to guess, but is easy to add for correctness.
      timingSafeEqual(projectDetailsPeerId, peerId) &&
      timingSafeEqual(inviteId, details.inviteId)
    if (!isDetailsForThisInvite) return
    rpc.off('got-project-details', onProjectDetails)
    resolve(details)
  })
})
```

Reply:
I think it needs to work this way because we want to remove the listener if the accept fails for any reason. I thought I'd done that but I guess not! Done in f71351d14c6c7ca297bf8f2846c9076797b3e3b3.